### ISSUE TITLE
Discuss new design

### ISSUE THREAD

**Issue by pchiusano** [scalaz/scalaz-stream#431]
#383 thread was getting really long, starting a fresh thread to discuss new design and it's implementation in the topic/0.8-redesign branch, which is badly named since there will be a 0.8 release without the new design

**Comment by pchiusano** [comment]
In reply to Pavel's comments [here](https://github.com/scalaz/scalaz-stream/issues/383#issuecomment-133881924), a couple notes, I can respond in more detail this week:

 The new implementation doesn't depend on Task directly anywhere, everything is defined in terms of the Async interface, so it should be easy to make whatever changes we want to Task. I just started by porting the scalaz Task and removing all the scalaz dependencies. I dont think I changed anything else. I kind of agree about retry and retryAccumulating, the question is whether we care about making Task a drop in replacement for the scalaz one. Maybe that is not really something we should worry too much about, since people can continue to use those Task by supplying the Async instance for it.

Hav some other comments will write up when in front of a computer.

**Comment by pchiusano** [comment]
Hey @pchlupacek sorry for the delay, here are some more detailed comments:

> on the runAsync of the Task. I just would like to double check we are not going to implement functionality that allows us to peek whether the task has been interrupted from the cb. I think this is usefull and to me was one of the limitiation of concurrent task. If yes, we will need to touch cb signature to return perhaps boolean instead of Unit.

I don't think this is necessary, especially since we are moving to a model in which a `Task` is always an atomic, uninterruptible thing. In fact, I think we should get rid of the `*-interruptibly` methods.

> overall async signatures of Task will be subject of SOE. Curious if we must accept that.

Can you elaborate? I do think we need a version of async which just runs the callback in the same thread, but yeah, it is not always stack safe. We could introduce `forkAsync` which has the same signature but runs the callback in a thread provided by a `Strategy`?

> do we really want retryAccumulating and retry on Task? I would rather to see that on Process. To me this seems like a bit of complex functionality to bake it in Task directly?

See my earlier comment. I think I'm in favor of getting rid of this. The `Task` in this library can be a minimal one, and people can use the scalaz one or write their own helper functions if they need something fancier or need additional functions.

> I am curious if Task.race wouldn't SOE at deeper task combinations. Also I have a feeling that we may have a questionable performance if we define chooseAny in terms of race for let's say 10k Tasks?

That sounds like a good test case. :) `chooseAny` and some other stuff probably deserve specialized implementations.

> I would suggest to have Sequential and Concurrent strategy marked traits (ie: StrategySequential, StrategyConcurrent). That, will allow Actor to switch based on whether there is fresh stack to more effective implementations, and potentially implementations, that may overcome SOE when combined together with Task.async. 

That could be a good idea. Want to play around with that and write up your findings?

> Anyhow do you have some plan on testing all that? Or just can we make a sort of plan to split testing between all of us?

Let me get all the basic stuff at least seemingly working then we can get organized on a test plan and divide up the work. I am still fixing major bugs. :) We can probably reuse a lot of the old tests.

Latest update: the stream functions seem to be working, I've written some torture tests, and performance is good. I'm working on a bug with the implementation of stream transducers, but after I get the transducers working I think I'll feel pretty good about the new implementation. I still have to do a writeup of the new design.

**Comment by pchlupacek** [comment]
@pchiusano thanks for explanation
> I don't think this is necessary, especially since we are moving to a model in which a Task is always an atomic, uninterruptible thing. In fact, I think we should get rid of the *-interruptibly methods.

Agreed, lets remove them. We never used them in user code anyway. 

>Can you elaborate? I do think we need a version of async which just runs the callback in the same thread, but yeah, it is not always stack safe. We could introduce forkAsync which has the same signature but runs the callback in a thread provided by a Strategy?

My issue is with this sort of code which I still see people writing.

```scala

 Task.async {  cb0 =>  Task.async { cb1 =>  ...... }  }
 // instead of 
 Task.async {}.flatMap { a => Task.async { .... }}

``` 
People do not write this of course directly, but hence `cb` is just function, they sometimes pass it around and forget about stack alltogether. I just would like to see if we can make this async callback a bit safer than it is. Let me look on `Async` signature, if we cannot start from there. 

>See my earlier comment. I think I'm in favor of getting rid of this. The Task in this library can be a minimal one, and people can use the scalaz one or write their own helper functions if they need something fancier or need additional functions.
 
Yes agreed.

> That could be a good idea. Want to play around with that and write up your findings?

Will take look on that

**Comment by pchiusano** [comment]
I've been making really good progress on this, to the point where the new design is actually:

* Pretty nice to use! I like it a lot, and it's much more expressive than the current library.
* Fast. Chunked awaits + pushback are a huge win. @djspiewak remember we discussed `take` and how it had to be un-chunky? Well it can now be implemented in a chunky fashion. If the last chunk reads too many elements, they just get pushed back into the `Handle`. Easy!
* _Working_, as far as I can tell, and I have some tests to prove it. Lots more testing needed, though, biggest area is around all the `Async` stuff. But the implementation is the sort of thing that, if it typechecks, can be buggy in only 15 possible ways and I've now gone through about 7 of those ways.... :) My suspicion is that if further bugs are found, the implementation will very rapidly stabilize.

I started working on a writeup of the new design, which also doubles as a user guide that I'd like us to keep updated:

https://github.com/scalaz/scalaz-stream/blob/topic/0.8-redesign/docs/guide.markdown 

It is a little introductory, but it covers all the major pieces. The links to the code in the doc don't work, btw. I'll fix that later.

I really am super excited about the new design but I would like to get feedback from folks. Do you like it? Do you have questions about how stuff will work? We can discuss all that stuff on this thread, then assuming we are feeling good, I'll kick of a new issue to start discussing the best migration path and getting organized on all the outstanding todos before we can do a release. I have a lot of questions about the best migration path. A lot of things will stay source compatible, but the biggest source-level change is in how transducers like `Process1`, `Tee`, and `Wye` are defined. 

I also would like to discuss whether we should narrow the scope of the core library. (Do we still want to include all of `async` and `io`, for instance?)

Also, I figured out [how to get Scala to not infer useless lubs like `Any` or `Serializable`]( https://github.com/scalaz/scalaz-stream/blob/topic/0.8-redesign/docs/guide.markdown#appendix-a1-sane-subtyping-with-better-error-messages) and am using this technique everywhere that covariance of `Stream` or `Pull` demands you lub a type parameter. It's stunning. Almost makes me not hate subtyping, ha. (/cc @mpilquist thanks for your help the other day).

Random question: the code does not seem to compile with 2.10.4. I'm guessing different compiler bugs. Unless these are super simple to deal with, I am fine just targeting 2.11+.

Okay everyone, have a good weekend!

Review by @djspiewak @runarorama @alissapajer @fthomas

**Comment by pchiusano** [comment]
Closing this
